### PR TITLE
consider label bounds when culling offscreen nodes

### PR DESCRIPTION
**Pull request by mggower**

### CODE REVIEW

File: src/renderers/webgl/objects/label/background.ts [hunk 1]
```diff
@@ -14,15 +14,15 @@ export class LabelBackground {
...
+  private rect: Size
   private _style: LabelBackgroundStyle
 
   constructor(container: Container, label: Text | BitmapText, style: LabelBackgroundStyle) {
     this.label = label
     this.container = container
     this._style = style
 
-    this.rect = { ...this.label.getLocalBounds() }
+    this.rect = { width: this.label.width, height: this.label.height }
```
Comment (mggower) [inline src/renderers/webgl/objects/label/background.ts:25]: getting very wonky results when calling `label.getBounds(true)` or `label.getLocalBounds()` with `BitmapText` objects - this seems to be the safer route

File: src/renderers/webgl/objects/label/background.ts [hunk 2]
```diff
@@ -90,6 +90,10 @@ export class LabelBackground {
     return undefined
   }
 
+  getBounds() {
+    return utils.getBounds(this.x ?? 0, this.y ?? 0, this.sprite.width, this.sprite.height, this.sprite.anchor.clone())
```
Comment (jameslaneconkling) [inline src/renderers/webgl/objects/label/background.ts:94]: why clone the anchor?

File: src/renderers/webgl/objects/label/background.ts [hunk 3]
```diff
@@ -14,15 +14,15 @@ export class LabelBackground {
   private sprite: Sprite
   private label: Text | BitmapText
   private container: Container
-  private rect: Rectangle
+  private rect: Size
```
Comment (jameslaneconkling) [inline src/renderers/webgl/objects/label/background.ts:17]: could we avoid allocating the object if we just set `private labelWidth: number`/`private labelHeight: number`?

**Review by jameslaneconkling — COMMENTED** [review]
two things to confirm:
* bounds are correctly recomputed when text changes
* perf impact: easiest way to do this is to initialize the test graph at or just above the min zoom needed to render labels and measure the fps when dragging the viewport around in circles. iirc should be something like 15fps. and then compare the numbers with and without the change. should also do this w/ and w/o label backgrounds.

**Comment by mggower** [comment]
@jameslaneconkling got around to address the above comments. I addressed some perf issues so label bounds is not recomputed on each frame but after an update or being repositioned. 

**Perf test:** dragging the viewport around in circles at min-zoom
- with labels: averages 18/19 fps
- with label backgrounds: averages 15/16 fps

**Review by jameslaneconkling — CHANGES_REQUESTED** [review]
Because label/background bounds are absolute (meaning the coordinates are relative to the trellis coordinate plane, not relative to the object/node they belong to), every time you move the object/node, you have to recompute those bounds. If instead you computed the width/height dimensions, you'd only have to do that computation when the label/background is created or the content changes, but not when it moves. 

Then the `webgl/node.ts` `visible()` method would look something like
```ts
let left: number
let right: number
let top: number
let bottom: number

if (this.label) {
  left = this.x - Math.min(this.strokes.radius, this.label.bounds.left)
  right = this.x + Math.max(this.strokes.radius, this.label.bounds.right)
  top = this.y - Math.min(this.strokes.radius, this.label.bounds.top)
  bottom = this.y + Math.max(this.strokes.radius, this.label.bounds.bottom)
} else {
  left = this.x - this.strokes.radius
  right = this.x + this.strokes.radius
  top = this.y - this.strokes.radius
  bottom = this.y + this.strokes.radius
}
```

The result should be more efficient and might simplify the object implementations too.

**Review by jameslaneconkling — APPROVED** [review]

**Pull request merged**